plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hadoop/TrinoHadoopCatalog.java

```diff
@@ -32,202 +40,227 @@
+import static io.trino.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
 import static io.trino.spi.StandardErrorCode.NOT_SUPPORTED;
+import static io.trino.spi.StandardErrorCode.SCHEMA_ALREADY_EXISTS;
+import static io.trino.spi.StandardErrorCode.SCHEMA_NOT_EMPTY;
+import static io.trino.spi.StandardErrorCode.SCHEMA_NOT_FOUND;
 import static java.util.Locale.ENGLISH;
 import static java.util.Objects.requireNonNull;
-import static org.apache.iceberg.CatalogProperties.WAREHOUSE_LOCATION;
-import static org.apache.iceberg.CatalogUtil.loadCatalog;
+import static java.util.function.Function.identity;
 
 public class TrinoHadoopCatalog
         extends AbstractTrinoCatalog
 {
-    static {
-        HadoopNative.requireHadoopNative();
-    }
-
-    private static final Map<String, String> EMPTY_SESSION_MAP = ImmutableMap.of();
-    private static final String CATALOG_IMPL = HadoopCatalog.class.getName();
-
-    private final HdfsEnvironment hdfsEnvironment;
-    private final Map<String, String> catalogProperties;
-    private final Cache<String, Catalog> catalogCache;
+    private static final Joiner SLASH = Joiner.on("/");
+    private static final Splitter DOT = Splitter.on('.');
+    private static final int PER_QUERY_CACHES_SIZE = 1000;
+    private final TrinoFileSystem trinoFileSystem;
+    private final CatalogName catalogName;
     private final String warehouse;
+    private final IcebergTableOperationsProvider tableOperationsProvider;
+    private final Cache<SchemaTableName, TableMetadata> tableMetadataCache = EvictableCacheBuilder.newBuilder()
+            .maximumSize(PER_QUERY_CACHES_SIZE)
+            .build();
 
-    public TrinoHadoopCatalog(CatalogName catalogName, HdfsEnvironment hdfsEnvironment, TypeManager typeManager, IcebergTableOperationsProvider tableOperationsProvider, boolean useUniqueTableLocation, IcebergConfig config)
+    public TrinoHadoopCatalog(
+            CatalogName catalogName,
+            TypeManager typeManager,
+            ConnectorIdentity identity,
+            IcebergTableOperationsProvider tableOperationsProvider,
+            TrinoFileSystemFactory fileSystemFactory,
+            boolean useUniqueTableLocation,
+            IcebergConfig icebergConfig)
     {
-        super(catalogName, typeManager, tableOperationsProvider, useUniqueTableLocation);
-        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
-        this.catalogProperties = convertToCatalogProperties(config);
-        this.catalogCache = SafeCaches.buildNonEvictableCache(CacheBuilder.newBuilder().maximumSize(config.getCatalogCacheSize()));
-        this.warehouse = requireNonNull(config.getCatalogWarehouse(), "warehouse is null");
-    }
+        super(catalogName, typeManager, tableOperationsProvider, fileSystemFactory, useUniqueTableLocation);
+        this.warehouse = requireNonNull(icebergConfig.getCatalogWarehouse(), "warehouse is null");
```

Comment:
We should make sure `warehouse` ends with a trailing `/`. The Azure storage lib requires it for directories (which the warehouse path always will be) and fires a very confusing exception when it's not there.
```suggestion
        this.warehouse = icebergConfig.getCatalogWarehouse() + icebergConfig.getCatalogWarehouse().endsWith('/') ? "" : "/";
```

Reply:
Not so sure we have to have one... Everywhere we deal with the joiner takes care of adding one at the end of the warehouse location whereever we need to use it.